ext/standard/random.c

```diff
@@ -85,26 +90,74 @@ static int php_random_bytes(void *bytes, size_t size)
 	}
 #elif HAVE_DECL_ARC4RANDOM_BUF
 	arc4random_buf(bytes, size);
+#elif HAVE_DECL_GETRANDOM
+	/* Linux getrandom(2) syscall */
+	size_t read_bytes = 0;
+	size_t amount_to_read = 0;
+	ssize_t n;
+
+	/* Keep reading until we get enough entropy */
+	do {
+		amount_to_read = size - read_bytes;
+		/* Below, (bytes + read_bytes)  is pointer arithmetic.
+
+		   bytes   read_bytes  size
+		     |      |           |
+		    [#######=============] (we're going to write over the = region)
+		             \\\\\\\\\\\\\
+		              amount_to_read
+
+		*/
+		n = getrandom(bytes + read_bytes, amount_to_read, 0);
+
+		if (n == -1) {
+			if (errno == EINTR || errno == EAGAIN) {
+				/* Try again */
+				continue;
+			}
+			/*
+				If the syscall fails, we are doomed. The loop that calls
+				php_random_bytes should be terminated by the exception instead
+				of proceeding to demand more entropy.
+			*/
+			zend_throw_exception(zend_ce_exception, "Could not gather sufficient random data", errno);
+			return FAILURE;
+		}
+
+		read_bytes += (size_t) n;
+	} while (read_bytes < size);
 #else
 	int    fd = RANDOM_G(fd);
+	struct stat st;
 	size_t read_bytes = 0;
+	ssize_t n;
 
 	if (fd < 0) {
-#if HAVE_DEV_ARANDOM
-		fd = open("/dev/arandom", O_RDONLY);
-#elif HAVE_DEV_URANDOM
+#if HAVE_DEV_URANDOM
 		fd = open("/dev/urandom", O_RDONLY);
 #endif
 		if (fd < 0) {
 			zend_throw_exception(zend_ce_exception, "Cannot open source device", 0);
 			return FAILURE;
 		}
+		/* Does the file exist and is it a character device? */
+		if (fstat(fd, &st) == 0 && S_ISCHR(st.st_mode)) {
+			RANDOM_G(fd) = fd;
+		} else {
+			zend_throw_exception(zend_ce_exception, "Error reading from source device", 0);
+			return FAILURE;
+		}
 
-		RANDOM_G(fd) = fd;
+	} else {
+		/* Is this still a valid file handle? */
+		if (!(fstat(fd, &st) == 0 && S_ISCHR(st.st_mode))) {
+			zend_throw_exception(zend_ce_exception, "Error reading from source device", 0);
+			return FAILURE;
+		}
```

Comment:
Right, I remember reading a way to sanity check that a while back, but can't find the reference. I'll update the PR as soon as I find it again.